### PR TITLE
Calendar API changes so that methods accepting strings return strings

### PR DESCRIPTION
**Pull request by chipkent**
Suggested changes to the Calendar API so that methods that accept string inputs return strings.

### CODE REVIEW

File: engine/time/src/main/java/io/deephaven/time/calendar/BusinessCalendar.java [hunk 1]
```diff
@@ -1287,7 +1291,7 @@ public LocalDate[] businessDates(final LocalDate start, final LocalDate end) {
      * @throws InvalidDateException if the dates are not in the valid range
      * @throws DateTimeUtils.DateTimeParseException if the string cannot be parsed
      */
-    public LocalDate[] businessDates(final String start, final String end) {
+    public String[] businessDates(final String start, final String end) {
```
Comment (jjbrosnan) [inline engine/time/src/main/java/io/deephaven/time/calendar/BusinessCalendar.java:1294]: Every other `businessDate` overload returns a `LocalDate[]`. I don't think the one that takes string inputs needs to be special.

File: engine/time/src/main/java/io/deephaven/time/calendar/BusinessCalendar.java [hunk 2]
```diff
@@ -1357,14 +1361,16 @@ public LocalDate[] nonBusinessDates(final LocalDate start, final LocalDate end,
      * @throws InvalidDateException if the dates are not in the valid range
      * @throws DateTimeUtils.DateTimeParseException if the string cannot be parsed
      */
-    public LocalDate[] nonBusinessDates(final String start, final String end, final boolean startInclusive,
+    public String[] nonBusinessDates(final String start, final String end, final boolean startInclusive,
```
Comment (jjbrosnan) [inline engine/time/src/main/java/io/deephaven/time/calendar/BusinessCalendar.java:1364]: Same comment as the one on `businessDates`

File: engine/time/src/main/java/io/deephaven/time/calendar/BusinessCalendar.java [hunk 3]
```diff
@@ -1430,7 +1436,7 @@ public LocalDate[] nonBusinessDates(final LocalDate start, final LocalDate end)
      * @throws InvalidDateException if the dates are not in the valid range
      * @throws DateTimeUtils.DateTimeParseException if the string cannot be parsed
      */
-    public LocalDate[] nonBusinessDates(final String start, final String end) {
+    public String[] nonBusinessDates(final String start, final String end) {
```
Comment (jjbrosnan) [inline engine/time/src/main/java/io/deephaven/time/calendar/BusinessCalendar.java:1439]: Same comment as the one I made on `businessDates`

File: engine/time/src/main/java/io/deephaven/time/calendar/BusinessCalendar.java [hunk 4]
```diff
@@ -165,6 +165,7 @@ private YearData getYearData(final int year) {
 
     // region Getters
 
+    // TODO: should there be a string equivalent?
```
Comment (jjbrosnan) [inline engine/time/src/main/java/io/deephaven/time/calendar/BusinessCalendar.java:168]: I don't think there needs to be one

File: engine/time/src/main/java/io/deephaven/time/calendar/BusinessCalendar.java [hunk 5]
```diff
@@ -1214,14 +1216,16 @@ public LocalDate[] businessDates(final LocalDate start, final LocalDate end, fin
      * @throws InvalidDateException if the dates are not in the valid range
      * @throws DateTimeUtils.DateTimeParseException if the string cannot be parsed
      */
-    public LocalDate[] businessDates(final String start, final String end, final boolean startInclusive,
+    public String[] businessDates(final String start, final String end, final boolean startInclusive,
```
Comment (jjbrosnan) [inline engine/time/src/main/java/io/deephaven/time/calendar/BusinessCalendar.java:1219]: Same comment as on the other overload.

File: engine/time/src/main/java/io/deephaven/time/calendar/BusinessCalendar.java [hunk 6]
```diff
@@ -2091,6 +2099,7 @@ public ZonedDateTime minusNonBusinessDays(final ZonedDateTime time, final int da
         return plusNonBusinessDays(time, -days);
     }
 
+    // TODO: should there be a string equivalent?
```
Comment (jjbrosnan) [inline engine/time/src/main/java/io/deephaven/time/calendar/BusinessCalendar.java:2102]: Isn't there already one?

File: engine/time/src/main/java/io/deephaven/time/calendar/BusinessCalendar.java [hunk 7]
```diff
@@ -2108,6 +2117,7 @@ public LocalDate futureBusinessDate(final int days) {
         return plusBusinessDays(calendarDate(), days);
     }
 
+    // TODO: should there be a string equivalent?
```
Comment (jjbrosnan) [inline engine/time/src/main/java/io/deephaven/time/calendar/BusinessCalendar.java:2120]: Isn't there already one?

File: engine/time/src/main/java/io/deephaven/time/calendar/BusinessCalendar.java [hunk 8]
```diff
@@ -2125,6 +2135,7 @@ public LocalDate pastBusinessDate(final int days) {
         return minusBusinessDays(calendarDate(), days);
     }
 
+    // TODO: should there be a string equivalent?
```
Comment (jjbrosnan) [inline engine/time/src/main/java/io/deephaven/time/calendar/BusinessCalendar.java:2138]: Isn't there already one?

File: engine/time/src/main/java/io/deephaven/time/calendar/Calendar.java [hunk 9]
```diff
@@ -89,6 +90,7 @@ public String toString() {
 
     // region Now
 
+    // TODO: should there be a string equivalent?
```
Comment (jjbrosnan) [inline engine/time/src/main/java/io/deephaven/time/calendar/Calendar.java:93]: I don't think so

File: engine/time/src/main/java/io/deephaven/time/calendar/Calendar.java [hunk 10]
```diff
@@ -117,6 +119,7 @@ public DayOfWeek dayOfWeek(final LocalDate date) {
         return DateTimeUtils.dayOfWeek(date);
     }
 
+    // TODO: should this return a String?
```
Comment (jjbrosnan) [inline engine/time/src/main/java/io/deephaven/time/calendar/Calendar.java:122]: All of the overloads for this method return a `DayOfWeek`. The one that takes string as input doesn't have to be special. It should conform to the behavior the overloads have.

File: engine/time/src/main/java/io/deephaven/time/calendar/Calendar.java [hunk 11]
```diff
@@ -392,6 +397,7 @@ public ZonedDateTime minusDays(final ZonedDateTime time, final int days) {
         return plusDays(time, -days);
     }
 
+    // TODO: should there be a string equivalent?
```
Comment (jjbrosnan) [inline engine/time/src/main/java/io/deephaven/time/calendar/Calendar.java:400]: I don't think so.

File: engine/time/src/main/java/io/deephaven/time/calendar/Calendar.java [hunk 12]
```diff
@@ -403,6 +409,7 @@ public LocalDate futureDate(final int days) {
         return plusDays(calendarDate(), days);
     }
 
+    // TODO: should there be a string equivalent?
```
Comment (jjbrosnan) [inline engine/time/src/main/java/io/deephaven/time/calendar/Calendar.java:412]: I don't think so.

**Review by jjbrosnan — COMMENTED** [review]
There were four methods in `BusinessCalendar` that I noticed behavior differed between overloads:

`minusBusinessDays`
`minusNonBusinessDays`
`plusBusinessDays`
`plusNonBusinessDays`

Those four have 4 overloads each - in all but 1 overload, the method returns the same type that's passed in. The only one where that's not the case is when a String is passed in. In that case, a `LocalDate` is returned. Those are the only 4 methods I think should change so that the behavior is consistent across all overloads.

After glancing at `Calendar` as well, there are two methods where I think this should change in the same way:

`minusDays`
`plusDays`

File: engine/time/src/main/java/io/deephaven/time/calendar/BusinessCalendar.java (hunk 1 above)
Comment (chipkent) [reply engine/time/src/main/java/io/deephaven/time/calendar/BusinessCalendar.java:1294]: I assumed that String in implies String out.  If someone is working in strings, do you expect that they would not want this result in strings as well?

File: engine/time/src/main/java/io/deephaven/time/calendar/BusinessCalendar.java (hunk 2 above)
Comment (chipkent) [reply engine/time/src/main/java/io/deephaven/time/calendar/BusinessCalendar.java:1364]: I assumed that String in implies String out.  If someone is working in strings, do you expect that they would not want this result in strings as well?

File: engine/time/src/main/java/io/deephaven/time/calendar/BusinessCalendar.java (hunk 3 above)
Comment (chipkent) [reply engine/time/src/main/java/io/deephaven/time/calendar/BusinessCalendar.java:1439]: I assumed that String in implies String out.  If someone is working in strings, do you expect that they would not want this result in strings as well?

File: engine/time/src/main/java/io/deephaven/time/calendar/BusinessCalendar.java (hunk 5 above)
Comment (chipkent) [reply engine/time/src/main/java/io/deephaven/time/calendar/BusinessCalendar.java:1219]: I assumed that String in implies String out.  If someone is working in strings, do you expect that they would not want this result in strings as well?

File: engine/time/src/main/java/io/deephaven/time/calendar/BusinessCalendar.java (hunk 4 above)
Comment (rbasralian) [reply engine/time/src/main/java/io/deephaven/time/calendar/BusinessCalendar.java:168]: I agree; people can convert these to strings on their own, or stick with java.time types (as we'd generally recommend).

I don't strongly oppose a `firstValidDateStr()`; I just wouldn't add it proactively. We can always add it later if people ask for it.

File: engine/time/src/main/java/io/deephaven/time/calendar/BusinessCalendar.java (hunk 5 above)
Comment (rbasralian) [reply engine/time/src/main/java/io/deephaven/time/calendar/BusinessCalendar.java:1219]: I think `String in -> String out`, `LocalDate in -> LocalDate out` is the most user-friendly

**Review by rbasralian — COMMENTED** [review]
Looks good to me — `String in -> String out`, `LocalDate in -> LocalDate out`, and always `LocalDate` if there are no date arguments.

It does make me wonder if we should make `parseLocalDate()` accept nulls — it would be nice to be able to do `parseLocalDate(someFunctionResultOrJoinedOnColumnThatMayBeNull)` whenever you have a string date, without having to handle nulls manually. We have `parseLocalDateQuiet()` (and similar for other parse methods), but those also swallow format exceptions.

**Review by jjbrosnan — APPROVED** [review]

**Comment by deephaven-internal** [comment]
Labels indicate documentation is required. Issues for documentation have been opened:

Community: https://github.com/deephaven/deephaven.io/issues/3616

**Pull request merged**